pkg/gardenlet/bootstrap/certificate/certificate_rotation.go

```diff
@@ -0,0 +1,225 @@
+	return &Manager{
+		seedName:               seedName,
+		clientMap:              clientMap,
+		gardenClientConnection: gardenClientConnection,
+		logger:                 logger.NewFieldLogger(logger.Logger, "certificate-rotation", seedName).Logger,
+		now:                    time.Now,
+	}
+}
+
+// ScheduleCertificateRotation waits until the currently used Garden cluster client certificate approaches expiration.
+// Then requests a new certificate and stores the kubeconfig in a secret (`gardenClientConnection.kubeconfigSecret`) on the Seed.
+// the argument is context.Cancel function to cancel the context of the Gardenlet used for graceful termination after a successful certificate rotation.
+// when the new gardenlet pod is started, it uses the rotated certificate stored in the secret in the Seed cluster
+func (cr *Manager) ScheduleCertificateRotation(ctx context.Context, gardenletCancel context.CancelFunc) {
+	go wait.Until(func() {
+		if err := cr.waitForCertificateRotation(ctx); err != nil {
+			cr.logger.Errorf("waiting for the certificate rotation failed: %v", err)
+			return
+		}
+
+		err := retry.Until(ctx, certificateWaitTimeout, func(ctx context.Context) (bool, error) {
+			ctxWithTimeout, cancel := context.WithTimeout(context.Background(), certificateWaitTimeout)
+			defer cancel()
+
+			err := cr.rotateCertificate(ctxWithTimeout)
+			if err != nil {
+				cr.logger.Errorf("certificate rotation failed: %v", err)
+				return retry.MinorError(err)
+			}
+			return retry.Ok()
+		})
+		if err != nil {
+			cr.logger.Errorf("Encountered error when trying to rotate certificate: %v", err)
+			return
+		}
+
+		cr.logger.Info("Terminating Gardenlet after successful certificate rotation.")
+		gardenletCancel()
+	}, time.Second, ctx.Done())
+}
+
+func (cr *Manager) waitForCertificateRotation(ctx context.Context) error {
+	return waitForCertificateRotation(ctx, cr.logger, cr.now, cr.gardenClientConnection, cr.clientMap, cr.seedName)
+}
+
+// waitForCertificateRotation determines and waits for the certificate rotation deadline.
+// Reschedules the certificate rotation in case the underlying certificate expiration date has changed in the meanwhile.
+func waitForCertificateRotation(ctx context.Context, logger logrus.FieldLogger, now func() time.Time, gardenClientConnection *config.GardenClientConnection, clientMap clientmap.ClientMap, seedName string) error {
+	cert, err := getCurrentCertificate(ctx, gardenClientConnection, clientMap, seedName)
+	if err != nil {
+		return err
+	}
+	deadline := nextRotationDeadline(*cert)
+	logger.Infof("Certificate expiration is %v, rotation deadline is %v", cert.Leaf.NotAfter, deadline)
+
+	if sleepInterval := deadline.Sub(now()); sleepInterval > 0 {
+		logger.Infof("Waiting for next certificate rotation in %d days (%s)", int(sleepInterval.Hours()/24), sleepInterval.Round(time.Second).String())
+		timer := time.NewTimer(sleepInterval)
+		defer timer.Stop()
+
+		// block until certificate rotation
+		<-timer.C
```

Comment:
Probably it is best to not unnecessarily leak go routines on termination. So please block at max until the context is cancelled and make the calling function stop retrying when the context has been cancelled (`wait.PollImmediateUntil` is probably what we need there).
```suggestion
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(sleepInterval):
		}
```

Reply:
> please block at max until the context is cancelled

what do you mean - I am not sure if I understand.

> stop retrying when the context has been cancelled

I adjusted the coding as depicted above so that ctx.Err() is returned as soon as ctx.Done() is called.
The calling function runs [until ctx.Done() is called](https://github.com/gardener/gardener/blob/67833d9f9bbfd0b9a8ddd0cfffec6a455ead2586/pkg/gardenlet/bootstrap/certificate/certificate_rotation.go#L94) - so it will not retry the execution - that should be sufficient?

Reply:
> > please block at max until the context is cancelled
> 
> what do you mean - I am not sure if I understand.

Now the function will return, when `ctx` is cancelled, so the go routine will exit properly when the gardenlet is terminated 👍 



> The calling function runs [until ctx.Done() is called](https://github.com/gardener/gardener/blob/67833d9f9bbfd0b9a8ddd0cfffec6a455ead2586/pkg/gardenlet/bootstrap/certificate/certificate_rotation.go#L94) - so it will not retry the execution - that should be sufficient?

Right, looks good now.